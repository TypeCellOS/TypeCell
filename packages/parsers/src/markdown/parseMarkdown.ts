import remarkParse from "remark-parse";
import { Root } from "remark-parse/lib";
import remarkStringify from "remark-stringify";
import { unified } from "unified";
import { Cell, Document, Language } from "../models";

export function markdownRootToDocument(tree: Root): Document {
  const nodes = tree.children;
  const cells: Cell[] = [];

  for (const node of nodes) {
    if (node.type === "code") {
      const lang = node.lang;
      if (lang && lang !== "typescript" && lang !== "css") {
        throw new Error("invalid markdown code block language");
      }
      cells.push({
        language: (lang as Language) || "typescript",
        code: node.value,
      });
    } else {
      let lastCell = cells[cells.length - 1];
      if (!lastCell || lastCell.language !== "markdown") {
        lastCell = {
          language: "markdown",
          code: "",
        };
        cells.push(lastCell);
      }
      const root: Root = {
        type: "root",
        children: [node],
      };

      const md = unified().use(remarkStringify).stringify(root) as string;

      if (lastCell.code.length) {
        lastCell.code += "\n";
      }
      lastCell.code += md;
    }
  }

  return {
    cells,
  };
}

export function markdownToDocument(markdown: string): Document {
  const tree = unified().use(remarkParse).parse(markdown);
  return markdownRootToDocument(tree);
}
