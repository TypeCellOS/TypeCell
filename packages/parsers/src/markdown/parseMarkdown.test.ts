import * as glob from "fast-glob";
import * as fs from "fs";
import * as path from "path";
import { describe, expect, it } from "vitest";
import { markdownToDocument } from "./parseMarkdown";

const ROOT_DIR = path.resolve(__dirname, "../../../../");

describe("parseMarkdown", () => {
  const mds = glob.sync(
    path.join(ROOT_DIR, "shared/test-data/markdown/**/*.md")
  );
  mds.forEach((md) => {
    it("converts " + md, () => {
      const data = fs.readFileSync(md, "utf-8");
      const converted = markdownToDocument(data);
      expect(converted).toMatchSnapshot();
    });
  });
});
