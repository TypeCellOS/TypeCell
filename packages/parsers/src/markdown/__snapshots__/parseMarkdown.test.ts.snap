// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseMarkdown > converts plainMarkdown.md 1`] = `
{
  "cells": [
    {
      "code": "### CSS

Use CSS to easily style the output of other cells (those written in Markdown or TypeScript).

<div class=\\"redText\\">This text is red, styled by the CSS cell above.</div>

### TypeScript / JavaScript

hello
",
      "language": "markdown",
    },
  ],
}
`;

exports[`parseMarkdown > converts withCSSAndTS.md 1`] = `
{
  "cells": [
    {
      "code": "### CSS

Use CSS to easily style the output of other cells (those written in Markdown or TypeScript).
",
      "language": "markdown",
    },
    {
      "code": ".redText {
  color: red;
}",
      "language": "css",
    },
    {
      "code": "<div class=\\"redText\\">This text is red, styled by the CSS cell above.</div>

### TypeScript / JavaScript
",
      "language": "markdown",
    },
    {
      "code": "export let message = \\"This is a TypeScript cell\\";",
      "language": "typescript",
    },
  ],
}
`;

exports[`parseMarkdown > converts withId.md 1`] = `
{
  "cells": [
    {
      "code": "plain text
",
      "language": "markdown",
    },
    {
      "code": "export let x = 5;",
      "language": "typescript",
    },
  ],
}
`;

exports[`parseMarkdown > converts withLibraryAndContext.md 1`] = `
{
  "cells": [
    {
      "code": "plain text
",
      "language": "markdown",
    },
    {
      "code": "import * as lodash from \\"lodash\\";
export let x = lodash.sum([5, 1]);",
      "language": "typescript",
    },
    {
      "code": "this shows how to reuse x:
",
      "language": "markdown",
    },
    {
      "code": "export let y = $.x + 1;",
      "language": "typescript",
    },
    {
      "code": "rain confetti!
",
      "language": "markdown",
    },
    {
      "code": "import { create } from \\"canvas-confetti\\";

export const canvas = document.createElement(\\"canvas\\");
canvas.width = 200;
canvas.height = 100;

const myConfetti = create(canvas);

// Drop some confetty every 500ms
setInterval(() => myConfetti({ particleCount: 70, origin: { y: 0 } }), 500);",
      "language": "typescript",
    },
  ],
}
`;
